{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eb7957f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as an\n",
    "from scipy import cluster\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import permutations\n",
    "import random\n",
    "from scipy import cluster\n",
    "from scipy import stats\n",
    "import os\n",
    "import glob\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from matplotlib.pyplot import figure\n",
    "from collections import Counter\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49cc87a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to compute counts per 10,000\n",
    "def cpm(x):\n",
    "    s = sum(list(x[0]))\n",
    "    x[0] = [i*10000/s for i in list(x[0])]\n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81eec2e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in and filter the data\n",
    "\n",
    "#First, get the 1-1 orthologs from ENSEMBL\n",
    "orthos = pd.read_csv(\"mouse_rat_orthologs.txt\", sep = \"\\t\").dropna()\n",
    "dorth = {}\n",
    "d_eg = {}\n",
    "for index, row in orthos.iterrows():\n",
    "    if row[\"Mouse homology type\"] == \"ortholog_one2one\":\n",
    "        dorth[row[\"Gene stable ID\"]] = row[\"Mouse gene stable ID\"]\n",
    "        d_eg[row[\"Mouse gene stable ID\"]] = row[\"Mouse gene name\"] \n",
    "\n",
    "ids = {\"RM1\":0, \"RM2\":0, \"RM3\":0, \"MR1\":0, \"MR2\":0, \"MR3\":0, \"WT\":0}\n",
    "#ids = {\"MR1\":0}\n",
    "\n",
    "#For each sample\n",
    "for i in ids:\n",
    "    \n",
    "    #Read in the data\n",
    "    v1 = sc.read_10x_mtx(i + \"_Mouse_Rat_Raw\")\n",
    "    v1.obs[\"Sample\"] = np.repeat(i, v1.obs.shape[0])\n",
    "    v1_m = sc.read_10x_mtx(i + \"_Mouse_Raw\")\n",
    "    v1_m.obs[\"Sample\"] = np.repeat(i, v1_m.obs.shape[0])\n",
    "    v1_r = sc.read_10x_mtx(i + \"_Rat_Raw\")\n",
    "    v1_r.obs[\"Sample\"] = np.repeat(i, v1_r.obs.shape[0])\n",
    "    v1_m.obs[\"BCS\"] = [x + \"_\" + i for x in list(v1_m.obs.index)]\n",
    "    v1_r.obs[\"BCS\"] = [x + \"_\" + i for x in list(v1_r.obs.index)]\n",
    "    \n",
    "    #Save the right gene ID (mouse or rat)\n",
    "    mouse_keep = []\n",
    "    rat_keep = []\n",
    "    for index, row in v1.var.iterrows():\n",
    "        if row[\"gene_ids\"][0] == \"m\":\n",
    "            mouse_keep.append(row[\"gene_ids\"])\n",
    "        elif row[\"gene_ids\"][0] == \"r\":\n",
    "            rat_keep.append(row[\"gene_ids\"])\n",
    "    \n",
    "    #Do basid filtering of genes and cells, filter to only keep the right genes\n",
    "    sc.pp.filter_cells(v1, min_genes=5)\n",
    "    sc.pp.filter_genes(v1, min_cells=1)\n",
    "    v1 = v1.T\n",
    "    v1_rg = v1[v1.obs[\"gene_ids\"].isin(rat_keep)]\n",
    "    v1_rg = v1_rg.copy().T\n",
    "    v1_mg = v1[v1.obs[\"gene_ids\"].isin(mouse_keep)]\n",
    "    v1_mg = v1_mg.copy().T\n",
    "    v1 = v1.T\n",
    "    \n",
    "    #Compute total counts per cell per genome and mitochondrial counts\n",
    "    v1_mg.var['mt'] = v1_mg.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(v1_mg, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    v1_rg.var['mt'] = v1_rg.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "    sc.pp.calculate_qc_metrics(v1_rg, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    \n",
    "    #Assign cell barcodes to species\n",
    "    rat = v1_rg.obs\n",
    "    mouse = v1_mg.obs\n",
    "    d = {}\n",
    "    \n",
    "    #Compute the total counts aligned to the rat genome and the total counts aligned to the mouse genome\n",
    "    for index, row in rat.iterrows():\n",
    "        d[index + \"_\" + i] = [row[\"total_counts\"]+1]\n",
    "\n",
    "    for index, row in mouse.iterrows():\n",
    "        d[index + \"_\" + i].append(row[\"total_counts\"]+1)\n",
    "    \n",
    "    #Classify cells using the cutoff of 70% of reads needing to come from one species\n",
    "    rat = []\n",
    "    mouse = []\n",
    "    doublet = []\n",
    "    for key in d.keys():\n",
    "        ratio = d[key][0]/(d[key][1] + d[key][0])\n",
    "        if ratio > 0.7:\n",
    "            rat.append(key)\n",
    "        elif ratio < 0.3:\n",
    "            mouse.append(key)\n",
    "        else:\n",
    "            doublet.append(key)\n",
    "    \n",
    "    #Standard filtering from scanpy\n",
    "    sc.pp.filter_cells(v1_m, min_genes=200)\n",
    "    sc.pp.filter_cells(v1_r, min_genes=200)\n",
    "    \n",
    "    #Now we work with the data aligned to only a single species\n",
    "    \n",
    "    #Create a new anndata object with correct species assignments and put in dictionary\n",
    "    v1_m_filt = v1_m[v1_m.obs[\"BCS\"].isin(mouse)]\n",
    "    v1_m_filt.obs[\"Species\"] = np.repeat(\"Mouse\", len(list(v1_m_filt.obs.index)))\n",
    "    v1_r_filt = v1_r[v1_r.obs[\"BCS\"].isin(rat)]\n",
    "    v1_r_filt.obs[\"Species\"] = np.repeat(\"Rat\", len(list(v1_r_filt.obs.index)))\n",
    "    \n",
    "    #Filter to keep only 1-1 orthologs and reassign gene names appropriately\n",
    "    v1_m_filt.var[\"Gene Name\"] = v1_m_filt.var.index\n",
    "    v1_m_filt = v1_m_filt.T\n",
    "    v1_m_filt = v1_m_filt[v1_m_filt.obs[\"gene_ids\"].isin(list(dorth.values()))]\n",
    "    v1_m_filt = v1_m_filt.copy().T\n",
    "\n",
    "    v1_r_filt.var[\"Gene Name\"] = v1_r_filt.var.index\n",
    "    v1_r_filt = v1_r_filt.T\n",
    "    v1_r_filt = v1_r_filt[v1_r_filt.obs[\"gene_ids\"].isin(list(dorth.keys()))]\n",
    "    v1_r_filt = v1_r_filt.copy().T\n",
    "    \n",
    "    #Rename the rat genes to their mouse ortholog\n",
    "    new_names = []\n",
    "    for index, row in v1_r_filt.var.iterrows():\n",
    "        new_names.append(d_eg[dorth[row[\"gene_ids\"]]])\n",
    "    v1_r_filt.var[\"New Gene Name\"] = new_names\n",
    "    v1_r_filt.var.index = v1_r_filt.var[\"New Gene Name\"]\n",
    "    c = Counter(v1_r_filt.var.index)\n",
    "    keepr = []\n",
    "    for key in c.keys():\n",
    "        if c[key] == 1:\n",
    "            keepr.append(key)\n",
    "\n",
    "    c = Counter(v1_m_filt.var.index)\n",
    "    keepm = []\n",
    "    for key in c.keys():\n",
    "        if c[key] == 1:\n",
    "            keepm.append(key)\n",
    "    \n",
    "    #Filter the genes\n",
    "    keep = np.intersect1d(keepr, keepm)\n",
    "    v1_r_filt = v1_r_filt.T\n",
    "    v1_r_filt = v1_r_filt[v1_r_filt.obs[\"New Gene Name\"].isin(keep)]\n",
    "    v1_m_filt = v1_m_filt.T\n",
    "    v1_m_filt = v1_m_filt[v1_m_filt.obs[\"Gene Name\"].isin(keep)]\n",
    "\n",
    "    v1_m_filt = v1_m_filt.copy().T\n",
    "    v1_r_filt = v1_r_filt.copy().T\n",
    "    print(i, \"# of rat cells\", len(list(v1_r_filt.obs.index)), \"# of mouse cells\", len(list(v1_m_filt.obs.index)))\n",
    "\n",
    "    #Now that gene names are fixed, concatenate and assign in dictionary\n",
    "    v = v1_m_filt.concatenate([v1_r_filt])\n",
    "    ids[i] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b09f8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Concatenate all the anndata objects from each sample together\n",
    "ind = 0\n",
    "v = 0\n",
    "for i in ids.keys():\n",
    "    if ind:\n",
    "        v = v.concatenate([ids[i]])\n",
    "    else:\n",
    "        v = ids[i]\n",
    "        ind = 1\n",
    "\n",
    "#Filter to only genes on sex chroms\n",
    "sc.pp.filter_genes(v, min_cells=10)\n",
    "f = list(pd.read_csv(\"Mouse_Sex_Chr_Genes.txt\", sep = \"\\t\", header = None)[0])\n",
    "vs = v.T\n",
    "vs = vs[vs.obs[\"New Gene Name-1\"].isin(f)]\n",
    "vs = vs.copy().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dda92c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### NOT USED ###\n",
    "\n",
    "#For each sample, split into mouse and rat\n",
    "v_MR1 = vs[vs.obs[\"Sample\"].isin([\"MR1\"])]\n",
    "v_MR1_m = v_MR1[v_MR1.obs[\"Species\"].isin([\"Mouse\"])]\n",
    "v_MR1_r = v_MR1[v_MR1.obs[\"Species\"].isin([\"Rat\"])]\n",
    "\n",
    "v_RM1 = vs[vs.obs[\"Sample\"].isin([\"RM1\"])]\n",
    "v_RM1_m = v_RM1[v_RM1.obs[\"Species\"].isin([\"Mouse\"])]\n",
    "v_RM1_r = v_RM1[v_RM1.obs[\"Species\"].isin([\"Rat\"])]\n",
    "\n",
    "v_RM2 = vs[vs.obs[\"Sample\"].isin([\"RM2\"])]\n",
    "v_RM2_m = v_RM2[v_RM2.obs[\"Species\"].isin([\"Mouse\"])]\n",
    "v_RM2_r = v_RM2[v_RM2.obs[\"Species\"].isin([\"Rat\"])]\n",
    "\n",
    "v_WT = vs[vs.obs[\"Sample\"].isin([\"WT\"])]\n",
    "v_WT_m = v_WT[v_WT.obs[\"Species\"].isin([\"Mouse\"])]\n",
    "v_WT_r = v_WT[v_WT.obs[\"Species\"].isin([\"Rat\"])]\n",
    "\n",
    "#Write out the CPM for each gene on sex chromosomes to determine the sex of each sample\n",
    "c = 0\n",
    "for i in [v_MR1_m, v_MR1_r, v_RM1_m, v_RM1_r, v_RM2_r, v_RM2_m, v_WT_m, v_WT_r]:\n",
    "    print(i)\n",
    "    c += 1\n",
    "    sc.pp.filter_cells(i, min_genes=100)\n",
    "    z = pd.DataFrame(i.X.todense())\n",
    "    z.columns = i.var.index\n",
    "    z.index = i.obs.index\n",
    "    #z[\"Species\"] = i.obs[\"Species\"]\n",
    "    #z[\"Sample\"] = i.obs[\"Sample\"]\n",
    "    z = z.T\n",
    "    z = pd.DataFrame(z.sum(axis=1))\n",
    "    z[\"CPM\"] = cpm(pd.DataFrame(z.sum(axis=1)))\n",
    "    z.to_csv(\"Final/DetermineSex/Determine_Sex\" + str(c) + \".csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "148b682c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we can proceed with the rest of the analysis\n",
    "ind = 0\n",
    "v = 0\n",
    "for i in ids.keys():\n",
    "    if ind:\n",
    "        v = v.concatenate([ids[i]])\n",
    "    else:\n",
    "        v = ids[i]\n",
    "        ind = 1\n",
    "\n",
    "#Do basic filtering again\n",
    "sc.pp.filter_genes(v, min_cells=10)\n",
    "\n",
    "#### Doing without sex chromsomes ####\n",
    "f = list(pd.read_csv(\"Mouse_Sex_Chr_Genes.txt\", sep = \"\\t\", header = None)[0])\n",
    "v = v.T\n",
    "v = v[~v.obs[\"New Gene Name-1\"].isin(f)]\n",
    "v = v.copy().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5778cb0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save an initial file as a starting place\n",
    "v.write(\"Final/All_Cells_Start.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e7811f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Do further filtering and clustering\n",
    "v.var['mt'] = v.var_names.str.startswith('mt-')  # annotate the group of mitochondrial genes as 'mt'\n",
    "sc.pp.calculate_qc_metrics(v, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "#Remove low quality cells\n",
    "v = v[v.obs.n_genes_by_counts < 7500, :]\n",
    "v = v[v.obs.pct_counts_mt < 15, :]\n",
    "\n",
    "#Save raw data\n",
    "v.raw = v\n",
    "\n",
    "#Normalize, compute highly variable genes and PCA as recommended by scanpy\n",
    "sc.pp.normalize_total(v, target_sum=1e4)\n",
    "sc.pp.log1p(v)\n",
    "sc.pp.highly_variable_genes(v, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "sc.pl.highly_variable_genes(v)\n",
    "sc.tl.pca(v, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(v, log=True)\n",
    "\n",
    "#Use harmony to correct for batch effects\n",
    "sce.pp.harmony_integrate(v, key = 'Species', basis = \"X_pca\", max_iter_harmony = 20)\n",
    "sc.pp.neighbors(v, n_neighbors=10, n_pcs=40, use_rep = \"X_pca_harmony\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1a0cd68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Do UMAP stuff and leiden clustering\n",
    "#0, 3, 4, 5 is blood-derived\n",
    "#1 is neuronal\n",
    "#2 is connective\n",
    "sc.tl.leiden(v, resolution=0.1)\n",
    "sc.tl.umap(v)\n",
    "sc.pl.umap(v, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb604c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Classify as from blood, brain, or connective tissue\n",
    "from_blood = [0, 1, 4, 5, 6, 7]\n",
    "from_con = [3]\n",
    "from_brain = [2, 8]\n",
    "\n",
    "#Create sub-objects for each one\n",
    "v_blood = v[v.obs[\"leiden\"].isin([str(x) for x in from_blood])]\n",
    "v_con = v[v.obs[\"leiden\"].isin([str(x) for x in from_con])]\n",
    "v_brain = v[v.obs[\"leiden\"].isin([str(x) for x in from_brain])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7608dfa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reclustering brain data using same strategy as above\n",
    "sc.pp.highly_variable_genes(v_brain, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "sc.pl.highly_variable_genes(v_brain)\n",
    "sc.tl.pca(v_brain, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(v_brain, log=True)\n",
    "sce.pp.harmony_integrate(v_brain, key = 'Species', basis = \"X_pca\", max_iter_harmony = 20)\n",
    "sc.pp.neighbors(v_brain, n_neighbors=10, n_pcs=40, use_rep = \"X_pca_harmony\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d82949b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reclustering brain data using same strategy as above, continued\n",
    "sc.tl.leiden(v_brain, resolution=0.5)\n",
    "sc.tl.umap(v_brain)\n",
    "sc.pl.umap(v_brain, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_brain, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_brain, n_genes=25, sharey=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f6b5777",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### DESCRIPTION OF CELL TYPES AND MARKER GENES USED TO IDENTIFY THEM ###\n",
    "\n",
    "#0 Glutamatergic neurons\n",
    "#1 GABAergic Meis2+ neurons\n",
    "#2 GABAergic progenitors\n",
    "#3 MGE inhibitory neurons (Nkx2-1+)\n",
    "#4 Glutamatergic progenitors\n",
    "#5 Immature glutamatergic neurons\n",
    "#6 Spinal cord excitatory neurons (Hoxb8+)\n",
    "#7 Spinal cord inhibitory neurons (Hoxb8+)\n",
    "#8 Nr2f2+ excitatory, contains CR cells so needs to be subclustered\n",
    "#9 GABAergic neurons\n",
    "#10 Intermediate progenitors\n",
    "#11 Thalamic neurons (Tcf7l2+)\n",
    "#12 Spinal cord progenitors\n",
    "#13 I think these are olfactory bulb exctiatory neurons\n",
    "#14 A very small population of glial-like cells I think\n",
    "#15 No idea\n",
    "#16 Also no idea\n",
    "\n",
    "#8 was split into:\n",
    "#17 Nr2f2+, Nhlh2+ excitatory neurons recently exiting the cell cycle\n",
    "#18 Cajal-Retzius cells\n",
    "v_brain = sc.read(\"Final/Brain_Subclustered_Prelim.h5ad\")\n",
    "excit = [0, 5]\n",
    "inhib = [1, 3, 9]\n",
    "exc_prog = [4]\n",
    "inh_prog = [2]\n",
    "inter_prog = [10]\n",
    "spinal_excit = [6]\n",
    "spinal_inhib = [7]\n",
    "excit_other = [17]\n",
    "\n",
    "sc.pl.umap(v_brain, color=['leiden'])\n",
    "sc.pl.umap(v_brain, color=[\"Gad2\", \"Slc17a6\", \"Eomes\", \"Gli3\", \"Hoxb8\", \"Reln\", \"Nr2f2\", \"Ebf1\", \"Mki67\", \"Zfhx3\", \"Meis2\", \"Nkx2-1\"], use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047c2f63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subcluster the inhibitory neurons\n",
    "v_inhib = v_brain[v_brain.obs[\"leiden\"].isin([str(x) for x in inhib])].copy()\n",
    "sc.tl.leiden(v_inhib, resolution=0.35)\n",
    "sc.tl.umap(v_inhib)\n",
    "sc.pl.umap(v_inhib, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_inhib, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_inhib, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d982b4cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Eliminate 6 as it is likely thalamic interneurons\n",
    "#Group 2 and 5 together due to lack of distinguishing markers\n",
    "#Rest stay separate\n",
    "sc.pl.umap(v_inhib, color=['leiden'])\n",
    "sc.pl.umap(v_inhib, color=['Erbb4', 'Meis2', 'Zfhx3', 'Nkx2-1', 'Usp29', 'Zic1', 'Tenm3', 'Tcf7l2', 'Nfib', 'Adarb2', 'Peg3'])\n",
    "v_inhib.write(\"GABAergic_Forebrain.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdbf1a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subclustering 8 as it contains Cajal-Retzius cells\n",
    "v_cr = v_brain[v_brain.obs[\"leiden\"].isin([\"8\"])]\n",
    "sc.tl.leiden(v_cr, resolution=0.05)\n",
    "sc.tl.umap(v_cr)\n",
    "sc.pl.umap(v_cr, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_cr, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_cr, n_genes=25, sharey=False)\n",
    "\n",
    "#Add the new correct annotations\n",
    "d = {}\n",
    "for index, row in v_cr.obs.iterrows():\n",
    "    d[index] = str(int(row[\"leiden\"]) + 17)\n",
    "x = v_brain.obs\n",
    "out = []\n",
    "for index, row in x.iterrows():\n",
    "    if index in d.keys():\n",
    "        l = list(row)\n",
    "        row[\"leiden\"] = d[index]  \n",
    "    out.append(row)\n",
    "new_obs = pd.DataFrame(out)\n",
    "new_obs.index = x.index\n",
    "new_obs.columns = x.columns\n",
    "v_brain.obs = new_obs\n",
    "\n",
    "v_brain.write(\"Final/Brain_Subclustered_Prelim.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fec8c30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now let's process and subcluster the connective tissue\n",
    "#Reclustering connective data\n",
    "sc.pp.highly_variable_genes(v_con, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "sc.pl.highly_variable_genes(v_con)\n",
    "sc.tl.pca(v_con, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(v_con, log=True)\n",
    "sce.pp.harmony_integrate(v_con, key = 'Species', basis = \"X_pca\", max_iter_harmony = 20)\n",
    "sc.pp.neighbors(v_con, n_neighbors=10, n_pcs=40, use_rep = \"X_pca_harmony\")\n",
    "\n",
    "sc.tl.leiden(v_con, resolution=0.15)\n",
    "sc.tl.umap(v_con)\n",
    "sc.pl.umap(v_con, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_con, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_con, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e22850d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#0 mesenchyme\n",
    "#1 chondrocytes\n",
    "#2 smooth muscle\n",
    "#3 endothelial cells\n",
    "#4, 5, 6 too few cells\n",
    "sc.pl.umap(v_con, color=['leiden', 'Mki67', \"Col2a1\", \"Acan\", \"Sox9\", \"Ptn\", \"Perp\", \"Runx2\", \"Tnnt1\", \"Col4a1\", \"Gpc5\", \"Col1a1\", \"Lamc3\", \"Runx1\", \"Tgfb2\", \"Adamts9\", \"Postn\", \"Epha7\"], use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cca98fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The next few cells subcluster the mesenchymal and chondrocytes just as done for the connective tissue\n",
    "\n",
    "v_mes = v_con[v_con.obs[\"leiden\"].isin([\"0\"])]\n",
    "sc.tl.leiden(v_mes, resolution=0.25)\n",
    "sc.tl.umap(v_mes)\n",
    "sc.pl.umap(v_mes, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_mes, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_mes, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95cd7389",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(v_mes, color=['leiden', 'Mki67', \"Col2a1\", \"Twist2\", \"Isl1\", \"Bmp5\", \"Mecom\", \"Gli3\", \"Col3a1\"], use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f82566e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_chondro = v_con[v_con.obs[\"leiden\"].isin([\"1\"])]\n",
    "sc.tl.leiden(v_chondro, resolution=0.25)\n",
    "sc.tl.umap(v_chondro)\n",
    "sc.pl.umap(v_chondro, color=['leiden'])\n",
    "x = sc.tl.rank_genes_groups(v_chondro, 'leiden', method='wilcoxon', use_raw = False)\n",
    "sc.pl.rank_genes_groups(v_chondro, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5732e6a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(v_chondro, color=['leiden', 'Mki67', \"Col1a2\", \"Bmpr1b\", \"Mef2c\"], use_raw = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7af4b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save files for later\n",
    "v_mes.write(\"Final/Mesenchymal_Subclustered_Prelim.h5ad\")\n",
    "v_chondro.write(\"Final/Chondrocyte_Subclustered_Prelim.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
