{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec67b26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import re\n",
    "from scipy.stats import pearsonr,spearmanr,fisher_exact,binom_test\n",
    "#import rpy2.robjects as robjects\n",
    "import random\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import os\n",
    "import seaborn as sns\n",
    "#import gseapy as gs\n",
    "from scipy.stats import norm\n",
    "import gseapy as gs\n",
    "from collections import Counter\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3742a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Colors\n",
    "mouse = \"#F2C911\"\n",
    "rat = \"#65B0AC\"\n",
    "intrinsic = \"#F55F00\"\n",
    "extrinsic = \"#7D9AF4\"\n",
    "reinforcing = \"#9B00F5\"\n",
    "opposing = \"#F50901\"\n",
    "interaction = \"#1E771A\"\n",
    "\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")\n",
    "d_ct_abrev = {\"chondrocytes\":\"chondrocyte\", \"Chondrocyte\":\"chondrocyte\", \"Forebrain glutamatergic progenitors\":\"brain.glut.prog\", \"Forebrain GABAergic progenitors\":\"brain.GABA.prog\", \"Intermediate progenitors\":\"inter.prog\", \"Forebrain glutamatergic neurons\":\"brain.glut.neu\", \"Forebrain GABAergic neurons\":\"brain.GABA.neu\", \"Spinal GABAergic neurons\":\"spine.GABA.neu\", \"Spinal glutamatergic neurons\":\"spine.glut.neu\", \"Chondrocytes\":\"chondrocyte\", \"Mesenchyme 0\":\"mesen.0\", \"Mesenchyme 2\":\"mesen.2\", \"Mesenchyme cycling\":\"mesen.cyc\", \"Forebrain GABAergic neurons 0\":\"Toss\"}\n",
    "\n",
    "#Define function to convert file name to cell type\n",
    "def file_to_celltype(x):\n",
    "    x = x.split(\"/\")[-1]\n",
    "    x = x.replace(\"_GO_Molecular_Function_2023\", \"\").replace(\"_GO_Biological_Process_2023\", \"\").replace(\"Mesechyme\", \"Mesenchyme\").replace(\"Brain_Or_2010_Div_New4_NewNorm_\", \"\")\n",
    "    x = x.replace(\"Mesenchymal_Or_2010_Div_New4_NewNorm_\", \"\").replace(\"Chondrocyte_Or_2010_Div_New4_NewNorm_\", \"\").replace(\".csv\", \"\").replace(\"_\", \" \").replace(\" all\", \"\")\n",
    "    x = x.replace(\"Glutamatergic\", \"Forebrain glutamatergic\").replace(\"GABAergic\", \"Forebrain GABAergic\").replace(\"Spinal Forebrain\", \"Spinal\").replace(\"Chondrocytes\", \"Chondrocyte\")\n",
    "    return x\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "#Define function to summarize enrichments from GSEAPY\n",
    "def summarize_enrichment(fold, out_file):\n",
    "    #Summarize the enrichments\n",
    "    out = []\n",
    "    \n",
    "    #Iterate through the folder of enrichments\n",
    "    for folder in os.listdir(\"Final/\" + fold):\n",
    "        v = pd.read_csv(\"Final/\" + fold + \"/\" + folder + \"/\" + \"gseapy.gene_set.prerank.report.csv\")\n",
    "        for index, row in v.iterrows():\n",
    "            #keep anything nominall significant\n",
    "            if row[\"FDR q-val\"] < 0.25:\n",
    "                out.append([folder, row[\"Term\"], row[\"FDR q-val\"], row[\"NES\"], row[\"Lead_genes\"]])\n",
    "    \n",
    "    #Write out to out_file\n",
    "    df = pd.DataFrame(out)\n",
    "    df.columns = [\"Cluster_Category\", \"Term\", \"FDR\", \"NES\", \"Ledge genes\"]\n",
    "    df = df.sort_values(\"FDR\")\n",
    "    df.to_csv(\"Final/Summarized_Enrichments_\" + out_file + \".csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bf470de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in list of imprinted genes and make dict in next cell\n",
    "imp_conf_list = []\n",
    "o = open(\"mouse_imprinted_genes.txt\")\n",
    "for line in o:\n",
    "    l = line.replace(\"\\xa0\", \"\").replace(\"\\n\", \"\").split(\"\\t\")\n",
    "    if (\"Imprinted\" in l[3] or \"Predicted\" in l[3]) and \"Not\" not in l[3]:\n",
    "        imp_conf_list.append(l)\n",
    "df_imp_info = pd.DataFrame(imp_conf_list)\n",
    "df_imp_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c01aed",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_imp = {\"Imprinted\":list(df_imp_info[0])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00e23a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Aggregating by tissue to compute enrichment for imprinted genes in interaction divergence\n",
    "\n",
    "df = pd.DataFrame()\n",
    "ind = 1\n",
    "\n",
    "#Set tissue and kind of divergence\n",
    "keep = \"Brain\"\n",
    "divergence = \"Interaction\"\n",
    "rank = True\n",
    "\n",
    "#Iterate through the files\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    \n",
    "    #If we want to keep all of them regardless of tissue\n",
    "    if keep == \"All\":\n",
    "        v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "        v[\"Abs \" + divergence.lower()] = np.abs(v[divergence])\n",
    "        v = v[[\"Abs \" + divergence.lower()]]\n",
    "        v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "        if ind:\n",
    "            df = v\n",
    "            ind = 0\n",
    "        else:\n",
    "            df = df.join(v, how=\"outer\")\n",
    "    else:\n",
    "        #If we only want neuronal cell types\n",
    "        if keep == \"Brain\":\n",
    "            if \"Brain\" in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v[\"Abs \" + divergence.lower()] = np.abs(v[divergence])\n",
    "                v = v[[\"Abs \" + divergence.lower()]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "        #If we only want connective tissue cell types\n",
    "        elif keep == \"Conn\":\n",
    "            if \"Brain\" not in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v[\"Abs \" + divergence.lower()] = np.abs(v[divergence])\n",
    "                v = v[[\"Abs \" + divergence.lower()]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "out = []\n",
    "indices = []\n",
    "\n",
    "#Require a gene be measured in at least 5 cell types if all, at least 3 cell types if connective tissue or nervous only\n",
    "for index, row in df.iterrows():\n",
    "    if keep == \"All\":\n",
    "        if len(row.dropna()) > 4:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "    else:\n",
    "        if len(row.dropna()) > 2:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "df2 = pd.DataFrame(out)\n",
    "df2.index = indices\n",
    "\n",
    "#Do absolute divergence\n",
    "df2.columns = [\"Abs \" + divergence.lower()]\n",
    "df2 = df2.sort_values(\"Abs \" + divergence.lower(), ascending = False)\n",
    "df2[\"Gene\"] = df2.index\n",
    "ranking = df2[[\"Gene\", \"Abs interaction\"]]\n",
    "if rank:\n",
    "    #Do rank transformation of the absolute log fold-changes around the median\n",
    "    #For example, if the absolute log fold-changes were 4.3, 2.1, 0.9, 0.4, 0.2\n",
    "    #Then this would output 2, 1, 0, -1, -2\n",
    "    front_half = list(range(1, ranking.shape[0]//2+1))\n",
    "    front_half.sort(reverse=True)\n",
    "    back_half = [-x for x in range(1, ranking.shape[0]//2+1)]\n",
    "    if ranking.shape[0] % 2 == 0:\n",
    "        ranking[\"Abs interaction\"] = front_half + back_half\n",
    "    else:\n",
    "        ranking[\"Abs interaction\"] = front_half + [0] + back_half\n",
    "#Actually do enrichment\n",
    "gs.prerank(rnk=ranking, gene_sets=d_imp, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Imprinting_Rank/' + keep + \"_\" + \"Imprinting\", format='png', seed=6, min_size = 10, max_size = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "440bdc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make dict to split by expressed allele\n",
    "d_imp2 = {\"Paternal\":list(df_imp_info[df_imp_info[4].isin([\"Paternal\"])][0]), \"Maternal\":list(df_imp_info[df_imp_info[4].isin([\"Maternal\"])][0])}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c530864",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Largely a repeat of the above code, but this time for signed divergence\n",
    "df = pd.DataFrame()\n",
    "ind = 1\n",
    "keep = \"Brain\"\n",
    "divergence = \"Interaction\"\n",
    "rank = False\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if keep == \"All\":\n",
    "        v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "        v = v[[divergence]]\n",
    "        v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "        if ind:\n",
    "            df = v\n",
    "            ind = 0\n",
    "        else:\n",
    "            df = df.join(v, how=\"outer\")\n",
    "    else:\n",
    "        if keep == \"Brain\":\n",
    "            if \"Brain\" in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v = v[[divergence]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "        elif keep == \"Conn\":\n",
    "            if \"Brain\" not in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v = v[[divergence]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "out = []\n",
    "indices = []\n",
    "for index, row in df.iterrows():\n",
    "    if keep == \"All\":\n",
    "        if len(row.dropna()) > 4:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "    else:\n",
    "        if len(row.dropna()) > 2:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "df2 = pd.DataFrame(out)\n",
    "df2.index = indices\n",
    "df2.columns = [divergence]\n",
    "df2 = df2.sort_values(divergence, ascending = False)\n",
    "df2[\"Gene\"] = df2.index\n",
    "\n",
    "#Doing for signed divergence\n",
    "ranking = df2[[\"Gene\", divergence]]\n",
    "\n",
    "gs.prerank(rnk=ranking, gene_sets=d_imp2, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Imprinting_Signed/' + keep + \"_\" + \"Imprinting_PatMat\", format='png', seed=6, min_size = 10, max_size = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93a7e5b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### None of the below was used so is not commented ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a4ef8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Repeat for DR/HR to match the embryonic heart data, not used\n",
    "df = pd.DataFrame()\n",
    "ind = 1\n",
    "keep = \"Conn\"\n",
    "divergence = \"DR/HR\"\n",
    "rank = False\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if keep == \"All\":\n",
    "        v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "        v = v[[divergence]]\n",
    "        v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "        if ind:\n",
    "            df = v\n",
    "            ind = 0\n",
    "        else:\n",
    "            df = df.join(v, how=\"outer\")\n",
    "    else:\n",
    "        if keep == \"Brain\":\n",
    "            if \"Brain\" in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v = v[[divergence]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "        elif keep == \"Conn\":\n",
    "            if \"Brain\" not in file:\n",
    "                v = pd.read_csv(\"Final/Div/\" + file).set_index(\"Unnamed: 0\")\n",
    "                v = v[[divergence]]\n",
    "                v.columns = [d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")]\n",
    "                if ind:\n",
    "                    df = v\n",
    "                    ind = 0\n",
    "                else:\n",
    "                    df = df.join(v, how=\"outer\")\n",
    "out = []\n",
    "indices = []\n",
    "for index, row in df.iterrows():\n",
    "    if keep == \"All\":\n",
    "        if len(row.dropna()) > 4:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "    else:\n",
    "        if len(row.dropna()) > 2:\n",
    "            out.append(np.mean(row.dropna()))\n",
    "            indices.append(index)\n",
    "df2 = pd.DataFrame(out)\n",
    "df2.index = indices\n",
    "df2.columns = [divergence]\n",
    "df2 = df2.sort_values(divergence, ascending = False)\n",
    "df2[\"Gene\"] = df2.index\n",
    "ranking = df2[[\"Gene\", divergence]]\n",
    "if rank:\n",
    "    front_half = list(range(1, ranking.shape[0]//2+1))\n",
    "    front_half.sort(reverse=True)\n",
    "    back_half = [-x for x in range(1, ranking.shape[0]//2+1)]\n",
    "    if ranking.shape[0] % 2 == 0:\n",
    "        ranking[\"Abs interaction\"] = front_half + back_half\n",
    "    else:\n",
    "        ranking[\"Abs interaction\"] = front_half + [0] + back_half\n",
    "gs.prerank(rnk=ranking, gene_sets=d_imp2, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Imprinting_Signed/' + keep + \"_\" + \"Imprinting_PatMat_HRDR\", format='png', seed=6, min_size = 10, max_size = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c45c3298",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot different clusters (in the genome) of imprinted genes\n",
    "\n",
    "file = \"Brain_Or_2010_Div_New4_NewNorm_GABAergic_progenitors.csv\"\n",
    "v = pd.read_csv(\"Final/Div/\" + file, sep = \",\").set_index(\"Unnamed: 0\")\n",
    "#vv = pd.read_csv(\"Final/NoDiv/\" + file.replace(\"Div\", \"NoDiv\"), sep = \",\").set_index(\"Unnamed: 0\")\n",
    "#v = pd.concat([v, vv])\n",
    "df_imp_info.index = df_imp_info[0]\n",
    "v_imp = v.join(df_imp_info).dropna()\n",
    "\n",
    "c1 = [\"Blcap, Nnat\", \"Zfp64\"]\n",
    "c2 = [\"Peg10\", \"Sgce\", \"Ppp1r9a\", \"Asb4\"]\n",
    "c3 = [\"Ndn\", \"Magel2\"]\n",
    "c4 = [\"Peg3\", \"Zim1\", \"Usp29\"]\n",
    "c5 = [\"Cd81\", \"Tssc4\", \"Kcnq1\", \"Cdkn1c\", \"Phlda2\", \"Nap1l4\"]\n",
    "c6 = [\"Grb10\", \"Ddc\"]\n",
    "add_on = []\n",
    "for index, row in v_imp.iterrows():\n",
    "    if index in c1:\n",
    "        add_on.append(\"Cluster 1\")\n",
    "    elif index in c2:\n",
    "        add_on.append(\"Cluster 2\")\n",
    "    elif index in c3:\n",
    "        add_on.append(\"Cluster 3\")\n",
    "    elif index in c4:\n",
    "        add_on.append(\"Cluster 4\")\n",
    "    elif index in c5:\n",
    "        add_on.append(\"Cluster 5\")\n",
    "    elif index in c6:\n",
    "        add_on.append(\"Cluster 6\")\n",
    "    else:\n",
    "        add_on.append(\"No cluster\")\n",
    "v_imp[\"Cluster\"] = add_on\n",
    "sns.scatterplot(data=v_imp, x=\"DR/HR\", y = \"DM/HM\", hue = \"Cluster\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
