{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a5d2d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import re\n",
    "from scipy.stats import pearsonr,spearmanr,fisher_exact,binom_test\n",
    "#import rpy2.robjects as robjects\n",
    "import random\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import os\n",
    "import seaborn as sns\n",
    "#import gseapy as gs\n",
    "from scipy.stats import norm\n",
    "import gseapy as gs\n",
    "from collections import Counter\n",
    "import scanpy as sc\n",
    "\n",
    "d_ct_abrev = {\"Chondrocyte\":\"chondrocyte\", \"Forebrain glutamatergic progenitors\":\"brain.glut.prog\", \"Forebrain GABAergic progenitors\":\"brain.GABA.prog\", \"Intermediate progenitors\":\"inter.prog\", \"Forebrain glutamatergic neurons\":\"brain.glut.neu\", \"Forebrain GABAergic neurons\":\"brain.GABA.neu\", \"Spinal GABAergic neurons\":\"spine.GABA.neu\", \"Spinal glutamatergic neurons\":\"spine.glut.neu\", \"Chondrocytes\":\"chondrocyte\", \"Mesenchyme 0\":\"mesen.0\", \"Mesenchyme 2\":\"mesen.2\", \"Mesenchyme cycling\":\"mesen.cyc\", \"Forebrain GABAergic neurons 0\":\"Toss\"}\n",
    "\n",
    "#Function to convert file to cell type\n",
    "def file_to_celltype(x):\n",
    "    x = x.split(\"/\")[-1]\n",
    "    x = x.replace(\"_GO_Molecular_Function_2023\", \"\").replace(\"_GO_Biological_Process_2023\", \"\").replace(\"Mesechyme\", \"Mesenchyme\").replace(\"Brain_Or_2010_Div_New4_NewNorm_\", \"\")\n",
    "    x = x.replace(\"Mesenchymal_Or_2010_Div_New4_NewNorm_\", \"\").replace(\"Chondrocyte_Or_2010_Div_New4_NewNorm_\", \"\").replace(\".csv\", \"\").replace(\"_\", \" \").replace(\" all\", \"\")\n",
    "    x = x.replace(\"Glutamatergic\", \"Forebrain glutamatergic\").replace(\"GABAergic\", \"Forebrain GABAergic\").replace(\"Spinal Forebrain\", \"Spinal\").replace(\"Chondrocytes\", \"Chondrocyte\")\n",
    "    return x\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23beb37b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gseapy import Biomart\n",
    "bm = Biomart()\n",
    "#Convert mouse gene names to human gene names with this\n",
    "m2h = bm.query(dataset='mmusculus_gene_ensembl',\n",
    "               attributes=['ensembl_gene_id','external_gene_name',\n",
    "                           'hsapiens_homolog_ensembl_gene',\n",
    "                           'hsapiens_homolog_associated_gene_name'])\n",
    "\n",
    "h2m = bm.query(dataset='hsapiens_gene_ensembl',\n",
    "               attributes=['ensembl_gene_id','external_gene_name',\n",
    "                           'mmusculus_homolog_ensembl_gene',\n",
    "                           'mmusculus_homolog_associated_gene_name'])\n",
    "m2h.to_csv(\"Mouse_To_Human_Gene_Conversions.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b51f8b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "m2h = pd.read_csv(\"Mouse_To_Human_Gene_Conversions.csv\").dropna()\n",
    "d = {}\n",
    "for index, row in m2h.iterrows():\n",
    "    d[row[\"external_gene_name\"]] = row[\"hsapiens_homolog_associated_gene_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a72480f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute number of CREs per gene\n",
    "v = pd.read_csv(\"Final/Human_Promoters_Ortho_Sorted_hg38_cCREs.sort.bed\", sep = \"\\t\", header = None)\n",
    "v = v[~v[5].isin([\"CTCF-only,CTCF-bound\"])]\n",
    "\n",
    "#Only count within 100 kb of the TSS\n",
    "v = v[v[10] < 100000]\n",
    "cres_per_gene = Counter(v[9])\n",
    "\n",
    "o = []\n",
    "for key in cres_per_gene.keys():\n",
    "    o.append([key, cres_per_gene[key]])\n",
    "cpg = pd.DataFrame(o)\n",
    "cpg.sort_values(1)\n",
    "cpg.columns = [\"Gene\", \"Number of CREs\"]\n",
    "cpg = cpg.set_index(\"Gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efea1166",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the widest possible interval for genes with multiple promoters\n",
    "v = pd.read_csv(\"Final/Human_Promoters_Ortho_Sorted_hg38.sort.bed\", sep = \"\\t\", header = None)\n",
    "o = []\n",
    "for key in list(set(list(v[3]))):\n",
    "    v2 = v[v[3].isin([key])]\n",
    "    for index, row in v2.iterrows():\n",
    "        pass\n",
    "    o.append([row[0], min(v2[1]), max(v2[2]), key])\n",
    "rns1 = pd.DataFrame(o)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42588223",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Write out if desired\n",
    "rns1.to_csv(\"Final/Human_Promoters_Ortho_Sorted_hg38_single.bed\", sep = \"\\t\", header = None, index = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97418c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sorted with sort -k1,1 -k2,2n\n",
    "v = pd.read_csv(\"Final/Human_Promoters_Ortho_Sorted_hg38_single.sort.bed\", sep = \"\\t\", header = None)\n",
    "o = []\n",
    "for chrom in [\"chr\" + str(x) for x in range(1, 23)] + [\"chrX\", \"chrY\"]:\n",
    "    \n",
    "    #Restrict to chrom\n",
    "    v_chrom = v[v[0].isin([chrom])]\n",
    "    genes = list(v_chrom[3])\n",
    "    for i in range(len(genes)):\n",
    "        if i == 0:\n",
    "            pass\n",
    "        elif i == len(genes) - 1:\n",
    "            pass\n",
    "        else:\n",
    "            #Get the distance to the nearest TSS on either side\n",
    "            cur_gene = v_chrom[v_chrom[3].isin([genes[i]])]\n",
    "            prev_val = int(v_chrom[v_chrom[3].isin([genes[i-1]])][2])\n",
    "            next_val = int(v_chrom[v_chrom[3].isin([genes[i+1]])][1])\n",
    "            o.append([genes[i], int(cur_gene[1])-prev_val + next_val - int(cur_gene[2])])\n",
    "rns = pd.DataFrame(o)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8024d8e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute Tau, a measure of tissue-specificity for our data\n",
    "def tau(x):\n",
    "    max_x = max(x)\n",
    "    n = len(x)\n",
    "    return np.sum((1-x/max_x)/(n-1))\n",
    "\n",
    "ind = 1\n",
    "df = 0\n",
    "\n",
    "#Go through pseudobulked unfiltered data\n",
    "for file in os.listdir(\"Final/Unfiltered_Pseudobulked/\"):\n",
    "    v = pd.read_csv(\"Final/Unfiltered_Pseudobulked/\" + file).set_index(\"Unnamed: 0\")\n",
    "    ind2 = 0\n",
    "    \n",
    "    #Compute mean normalized CPM\n",
    "    if \"Brain\" in file and \"neurons_0\" not in file:\n",
    "        v[\"Mean Norm CPM\"] = (v[\"MR1_mi Norm CPM\"] + v[\"MR1_ri Norm CPM\"] + v[\"RM1_mi Norm CPM\"] + v[\"RM1_ri Norm CPM\"])/4\n",
    "        v = v[[\"Mean Norm CPM\"]]\n",
    "        v.columns = [\"Mean norm CPM \" + file]\n",
    "        ind2 = 1\n",
    "    elif (\"Chondro\" in file or \"Mesenchyme\" in file) and \"Mesechyme_2\" not in file:\n",
    "        v[\"Mean Norm CPM\"] = (v[\"MR1_mi Norm CPM\"] + v[\"MR1_ri Norm CPM\"] + v[\"RM2_mi Norm CPM\"] + v[\"RM2_ri Norm CPM\"])/4\n",
    "        v = v[[\"Mean Norm CPM\"]]\n",
    "        v.columns = [\"Mean norm CPM \" + file]\n",
    "        ind2 = 1\n",
    "\n",
    "    if ind and ind2:\n",
    "        ind = 0\n",
    "        df = v\n",
    "    elif ind2:\n",
    "        df = df.join(v)\n",
    "\n",
    "#Restrict to only genes that are appreciabilty expressed in at least on cell type\n",
    "df[\"Max Exp\"] = np.max(df, axis = 1)\n",
    "df = df[df[\"Max Exp\"] > 0.2]\n",
    "\n",
    "#Compute tau\n",
    "df[\"Tau\"] = df.apply(tau, axis = 1)\n",
    "TAU = df[[\"Tau\"]]\n",
    "\n",
    "out = []\n",
    "for index, row in TAU.iterrows():\n",
    "    try:\n",
    "        out.append([d[index]] + list(row))\n",
    "    except:\n",
    "        pass\n",
    "TAU = pd.DataFrame(out).set_index(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64f3ce90",
   "metadata": {},
   "outputs": [],
   "source": [
    "rns.columns = [\"Gene\", \"Distance to nearest genes\"]\n",
    "rns = rns.set_index(\"Gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0070012",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Working with the Kaessmann files\n",
    "#Convert rat gene names to mouse and restrict to one-to-one orthos\n",
    "v_rat = pd.read_csv(\"Kaessmann_Bulk/Rat_rpkm.txt\", sep = \" \").set_index(\"Names\")\n",
    "orthos = pd.read_csv(\"mouse_rat_orthologs.txt\", sep = \"\\t\").dropna()\n",
    "orthos = orthos.set_index(\"Gene stable ID\")\n",
    "orthos = orthos[orthos[\"Mouse homology type\"] == \"ortholog_one2one\"]\n",
    "v_rat = v_rat.join(orthos).dropna()\n",
    "new_ind = []\n",
    "for i in list(v_rat.index):\n",
    "    new_ind.append(orthos.loc[i][\"Mouse gene name\"])\n",
    "v_rat.index = new_ind"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbe34b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Restrict to one-to-one orthos\n",
    "v_mouse = pd.read_csv(\"Kaessmann_Bulk/Mouse_rpkm.txt\", sep = \" \").set_index(\"Names\")\n",
    "orthos = pd.read_csv(\"mouse_rat_orthologs.txt\", sep = \"\\t\").dropna()\n",
    "orthos = orthos.set_index(\"Mouse gene stable ID\")\n",
    "orthos = orthos[orthos[\"Mouse homology type\"] == \"ortholog_one2one\"]\n",
    "v_mouse = v_mouse.join(orthos).dropna()\n",
    "new_ind = []\n",
    "for i in list(v_mouse.index):\n",
    "    new_ind.append(orthos.loc[i][\"Mouse gene name\"])\n",
    "v_mouse.index = new_ind"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "618ee40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get only genes measured in both\n",
    "keep = np.intersect1d(v_mouse.index, v_rat.index)\n",
    "v_rat = v_rat.loc[keep]\n",
    "v_mouse = v_mouse.loc[keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b69b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pull out the relevant embryonic time points for mouse\n",
    "mcol = np.array(v_mouse.columns).astype(str)\n",
    "keep11 = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e11.5\") != -1)\n",
    "keep13 = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e13.5\") != -1)\n",
    "keep15 = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e15.5\") != -1)\n",
    "out = []\n",
    "for index, row in v_mouse.iterrows():\n",
    "    data_11 = row[keep11]\n",
    "    data_13 = row[keep13]\n",
    "    data_15 = row[keep15]\n",
    "    labs = list(np.repeat(11.5, len(data_11))) + list(np.repeat(13.5, len(data_13))) + list(np.repeat(15.5, len(data_15)))\n",
    "    data_all = list(data_11) + list(data_13) + list(data_15)\n",
    "    #Compute correlation of expression with time\n",
    "    p = pearsonr(data_all, labs)\n",
    "    out.append([index, p[0], p[1]])\n",
    "df_mouse = pd.DataFrame(out)\n",
    "\n",
    "assert(list(v_mouse.index) == list(v_rat.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a468c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pull out the relevant embryonic time points for rat\n",
    "rcol = np.array(v_rat.columns).astype(str)\n",
    "keep11 = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e13\") != -1)\n",
    "keep13 = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e15\") != -1)\n",
    "keep15 = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e17\") != -1)\n",
    "out = []\n",
    "for index, row in v_rat.iterrows():\n",
    "    data_11 = row[keep11]\n",
    "    data_13 = row[keep13]\n",
    "    data_15 = row[keep15]\n",
    "    labs = list(np.repeat(11.5, len(data_11))) + list(np.repeat(13.5, len(data_13))) + list(np.repeat(15.5, len(data_15)))\n",
    "    data_all = list(data_11) + list(data_13) + list(data_15)\n",
    "    #Compute correlation of expression with time\n",
    "    p = pearsonr(data_all, labs)\n",
    "    out.append([index, p[0], p[1]])\n",
    "df_rat = pd.DataFrame(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d46fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Join rat and mouse\n",
    "df_rat.columns = [\"Gene\", \"Pearson r rat\", \"Pearson p-value rat\"]\n",
    "df_rat = df_rat.set_index(\"Gene\")\n",
    "df_mouse.columns = [\"Gene\", \"Pearson r mouse\", \"Pearson p-value mouse\"]\n",
    "df_mouse = df_mouse.set_index(\"Gene\")\n",
    "df = df_rat.join(df_mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab76f52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Multiply pearson rho values\n",
    "df[\"Mult\"] = df[\"Pearson r rat\"]*df[\"Pearson r mouse\"]\n",
    "df = df.sort_values(\"Mult\").dropna()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38dccc9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#If negative, then it is likely these genes have opposing expression trajectories in mouse and rat\n",
    "df_oppo = df[df[\"Mult\"] < -0.25]\n",
    "\n",
    "#Subset to those that are up in rat and down in mouse and vice versa\n",
    "upr_downm = df_oppo[(df_oppo[\"Pearson r rat\"] > 0) & (df_oppo[\"Pearson r mouse\"] < 0)]\n",
    "upm_downr = df_oppo[(df_oppo[\"Pearson r rat\"] < 0) & (df_oppo[\"Pearson r mouse\"] > 0)]\n",
    "d_DifDir = {\"DifMouseRat\":list(df_oppo.index)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34bec075",
   "metadata": {},
   "outputs": [],
   "source": [
    "#If there is a major role for changes in expression trajectory in driving our results, we would expect:\n",
    "#Enrichment of high interaction with opposite genes\n",
    "#We only see a weak enrichment in Spinal glutamatergic neurons, suggesting this isn't a major player\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v = v.sort_values(\"Proportion interaction\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Proportion interaction\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_DifDir, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Interaction_DifTimePattern/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9036c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "#If there is a major role for changes in expression trajectory in driving our results, we would expect:\n",
    "#Enrichment of UpRat_DownMouse at the bottom of the list\n",
    "#Enrichment of UpMouse_DownRat at the top of the list\n",
    "#We only see one of these things for three cell types\n",
    "#Combined with the above I don't think we have strong evidence that opposite expression trajectories are a likely cause of interactions\n",
    "\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v[\"Signed interaction\"] = np.sign(v[\"Interaction\"])*v[\"Proportion interaction\"]\n",
    "        v = v.sort_values(\"Signed interaction\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Signed interaction\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_DifDir, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Interaction_DifTimePatternSigned/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244d99ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get genes going in the same directions\n",
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "\n",
    "#Split into down in both and up in both\n",
    "up_both = df_same[df_same[\"Pearson r rat\"] > 0]\n",
    "down_both = df_same[df_same[\"Pearson r rat\"] < 0]\n",
    "d_SameDir = {\"SameDir\":list(df_same.index)}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61daf47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We would expect that same direction would be enriched for opposing/reinforcing\n",
    "#So we should see enrichment at the top of the list in this case\n",
    "#We see weak enrichment for Intermediate progenitors and spinal GABAergic neurons weaker enrichments in other cell types\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v[\"Multiply\"] = np.abs(v[\"Proportion intrinsic\"]*v[\"Proportion extrinsic\"])\n",
    "        v = v.sort_values(\"Multiply\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Multiply\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_SameDir, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_ReinfOppo_SameDir_Abs/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6715a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Renormalize all columns to sum to 1000000\n",
    "v_mouseT = v_mouse.T\n",
    "v_mouseT = v_mouseT.loc[np.setdiff1d(v_mouseT.index, [\"Gene stable ID\", \"Mouse gene name\", \"Mouse orthology confidence [0 low, 1 high]\", \"Mouse homology type\", \"Gene name\"])].copy()\n",
    "out = []\n",
    "for index, row in v_mouseT.iterrows():\n",
    "    out.append(row/np.sum(row)*1000000)\n",
    "v_mouse_new = pd.DataFrame(out).T\n",
    "v_mouse_new.columns = v_mouseT.index\n",
    "v_mouse_new.index = v_mouseT.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4e98b8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Do the same renormalization for rat\n",
    "v_ratT = v_rat.T\n",
    "v_ratT = v_ratT.loc[np.setdiff1d(v_ratT.index, [\"Gene stable ID\", \"rat gene name\", \"rat orthology confidence [0 low, 1 high]\", \"rat homology type\", \"Gene name\", 'Mouse gene name', 'Mouse gene stable ID', 'Mouse homology type', 'Mouse orthology confidence [0 low, 1 high]'])].copy()\n",
    "out = []\n",
    "for index, row in v_ratT.iterrows():\n",
    "    out.append(row/np.sum(row)*1000000)\n",
    "v_rat_new = pd.DataFrame(out).T\n",
    "v_rat_new.columns = v_ratT.index\n",
    "v_rat_new.index = v_ratT.columns\n",
    "v_rat_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63076b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Need to use the Kaessmann data to determine whether higher expression in mouse\n",
    "\n",
    "#Again get the right columns\n",
    "mcol = np.array(v_mouse_new.columns).astype(str)\n",
    "keep11m = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e11.5\") != -1)\n",
    "keep13m = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e13.5\") != -1)\n",
    "keep15m = np.flatnonzero(np.core.defchararray.find(mcol, \"Brain.e15.5\") != -1)\n",
    "\n",
    "rcol = np.array(v_rat_new.columns).astype(str)\n",
    "keep11r = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e13\") != -1)\n",
    "keep13r = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e15\") != -1)\n",
    "keep15r = np.flatnonzero(np.core.defchararray.find(rcol, \"Brain.e17\") != -1)\n",
    "\n",
    "high_rat = []\n",
    "high_mouse = []\n",
    "same_expr = []\n",
    "for index in list(v_mouse_new.index):\n",
    "    \n",
    "    #Get the expression levels at each stage for each sample\n",
    "    row_m = v_mouse_new.loc[index]\n",
    "    data_11m = row_m[keep11m]\n",
    "    data_13m = row_m[keep13m]\n",
    "    data_15m = row_m[keep15m]\n",
    "    \n",
    "    row_r = v_rat_new.loc[index]\n",
    "    data_11r = row_r[keep11r]\n",
    "    data_13r = row_r[keep13r]\n",
    "    data_15r = row_r[keep15r]\n",
    "    \n",
    "    #Compute the log fold-change between mouse and rat at equivalent stages\n",
    "    lfc13 = np.log2((1+np.mean(data_13m))/(1+np.mean(data_11r)))\n",
    "    lfc15 = np.log2((1+np.mean(data_15m))/(1+np.mean(data_13r)))\n",
    "    \n",
    "    #Compute the log fold-change at exactly the same timepoint\n",
    "    lfcNotStageMatch = np.log2((1+np.mean(data_13m))/(1+np.mean(data_13r)))\n",
    "    \n",
    "    #Split into genes with higher expression in mouse and higher expression in rat\n",
    "    if lfc13 > 0.5 and lfc15 > 0.5:\n",
    "        high_mouse.append(index)\n",
    "    elif lfc13 < -0.5 and lfc15 < -0.5:\n",
    "        high_rat.append(index)\n",
    "    #If we add the restriction of abs(lfcNotStageMatch) > 0.5 there are very few genes!\n",
    "    #Indicates that our extrinsic genes are probably fine\n",
    "    #Get genes with no difference at the same matched time point\n",
    "    elif abs(lfc13) < 0.25 and abs(lfc15) < 0.25:\n",
    "        same_expr.append(index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "777fb81a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "up_both = df_same[df_same[\"Pearson r rat\"] > 0]\n",
    "down_both = df_same[df_same[\"Pearson r rat\"] < 0]\n",
    "\n",
    "#Get genes that should show opposing intrinsic/extrinsic based on their expression trajectories\n",
    "should_be_oppo1 = np.intersect1d(high_mouse, up_both.index)\n",
    "should_be_oppo2 = np.intersect1d(high_rat, down_both.index)\n",
    "\n",
    "#Get genes that should show reinforcing intrinsic/extrinsic based on their expression trajectories\n",
    "should_be_reinf1 = np.intersect1d(high_rat, up_both.index)\n",
    "should_be_reinf2 = np.intersect1d(high_mouse, down_both.index)\n",
    "d_OR = {\"ShouldBeOppo\":list(should_be_oppo1) + list(should_be_oppo2), \"ShouldBeReinf\":list(should_be_reinf1) + list(should_be_reinf2)}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3358db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Signed test where opposite should be enriched for Mouse expr > Rat expr and increasing, Mouse expr < Rat expr and decreasing\n",
    "#Reinforcing should be enriched for Mouse expr > Rat expr and decreasing, Mouse expr < Rat expr and increasing\n",
    "#We don't see many enrichments, again implying that this is not a super major confounder for our results\n",
    "\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v.index = v[\"Unnamed: 0\"]\n",
    "        v[\"Multiply\"] = np.sign(v[\"Intrinsic\"])*v[\"Proportion intrinsic\"]*v[\"Proportion extrinsic\"]*np.sign(v[\"Extrinsic\"])\n",
    "                \n",
    "        v = v.sort_values(\"Multiply\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Multiply\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_OR, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_ReinfOppo_SameDirSigned/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32cb1caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "up_both = df_same[df_same[\"Pearson r rat\"] > 0]\n",
    "down_both = df_same[df_same[\"Pearson r rat\"] < 0]\n",
    "d_SameDir = {\"SameDir\":list(df_same.index)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81a2988a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We would expect that extrinsic would be enriched for same direction\n",
    "#So we should see enrichment at the top of the list in this case\n",
    "#We see weak enrichments in multiple cell types, indicating that this is not a major issue\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v = v.sort_values(\"Proportion extrinsic\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Proportion extrinsic\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_SameDir, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Extrinsic_SameDir/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "691697e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "up_both = df_same[df_same[\"Pearson r rat\"] > 0]\n",
    "down_both = df_same[df_same[\"Pearson r rat\"] < 0]\n",
    "d_SameDir = {\"UpBoth\":np.intersect1d(list(up_both.index), same_expr), \"DownBoth\":np.intersect1d(list(down_both.index), same_expr)}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edea1407",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We would expect an even stronger enrichment for those genes where mouse and rat have similar expression at E13.5 and E15.5\n",
    "#There are close to zero genes that have similar expression at E13.5 and E15.5 but a larger difference between stage matched in bulk data\n",
    "#This indicates off the bat that the extrinsic genes are probably truly extrinsic\n",
    "#Even with more relaxed criteria there are very few genes that have the same expression at E13.5 and E15.5 in mouse/rat (at least in bulk)\n",
    "#And are increasing/decreasing in expression over time\n",
    "#This suggests that the kind of heterochronic shift needed to explain extrinsic divergence is extremely rare at the bulk level\n",
    "#But that when it occurs it can indeed masquerade extrinsic divergence, at least in spinal neurons where we see enrichment\n",
    "\n",
    "\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"New4_NewNorm\" in file and \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v[\"Signed proportion extrinsic\"] = np.sign(v[\"Extrinsic\"])*v[\"Proportion extrinsic\"]\n",
    "        v = v.sort_values(\"Signed proportion extrinsic\", ascending = False)\n",
    "        ranking = v[[\"Unnamed: 0\", \"Signed proportion extrinsic\"]]\n",
    "        try:\n",
    "            gs.prerank(rnk=ranking, gene_sets=d_SameDir, threads=4, permutation_num=1000, outdir= 'C:/Users/astar/Chimerism/Final/GSEAPY_New4_Extrinsic_SameDirSameExpr/' + file.replace(\".csv\", \"\"), format='png', seed=6, min_size = 10, max_size = 30000)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21172b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Want to compute Tau and variance of the log fold-changes across tissues for the Kaessman data\n",
    "mcol = np.array(v_mouse_new.columns).astype(str)\n",
    "keep13 = np.flatnonzero(np.core.defchararray.find(mcol, \"e13.5\") != -1)\n",
    "keep13 = mcol[keep13]\n",
    "\n",
    "#Get the right stage for mouse\n",
    "v_mouse_new13 = v_mouse_new[keep13]\n",
    "mcols2 = np.array(v_mouse_new13.columns).astype(str)\n",
    "\n",
    "#Get the names of all tissues\n",
    "tissues = list(set([x.split(\".\")[0] for x in list(keep13)]))\n",
    "out = []\n",
    "for tissue in tissues:\n",
    "    #get the correct tissue and compute mean expression\n",
    "    keepT = np.flatnonzero(np.core.defchararray.find(mcols2, tissue) != -1)\n",
    "    keepT = mcols2[keepT]\n",
    "    v_mouse_new13T = v_mouse_new13[keepT]\n",
    "    out.append([tissue] + list(np.mean(v_mouse_new13T, axis = 1)))\n",
    "    \n",
    "#Create data frame\n",
    "df_m13 = pd.DataFrame(out)\n",
    "df_m13.columns = [\"Tissue\"] + list(v_mouse_new13.index)\n",
    "df_m13 = df_m13.set_index(\"Tissue\").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdbf138a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Want to compute Tau and variance of the log fold-changes across tissues for the Kaessman data, this time for rat\n",
    "#Repeated from code above\n",
    "rcol = np.array(v_rat_new.columns).astype(str)\n",
    "keep13 = np.flatnonzero(np.core.defchararray.find(rcol, \"e15\") != -1)\n",
    "keep13 = rcol[keep13]\n",
    "v_rat_new13 = v_rat_new[keep13]\n",
    "rcols2 = np.array(v_rat_new13.columns).astype(str)\n",
    "tissues = list(set([x.split(\".\")[0] for x in list(keep13)]))\n",
    "out = []\n",
    "for tissue in tissues:\n",
    "    keepT = np.flatnonzero(np.core.defchararray.find(rcols2, tissue) != -1)\n",
    "    keepT = rcols2[keepT]\n",
    "    v_rat_new13T = v_rat_new13[keepT]\n",
    "    out.append([tissue] + list(np.mean(v_rat_new13T, axis = 1)))\n",
    "df_r13 = pd.DataFrame(out)\n",
    "df_r13.columns = [\"Tissue\"] + list(v_rat_new13.index)\n",
    "df_r13 = df_r13.set_index(\"Tissue\").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "181d0c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Require at least moderate expression in one tissue\n",
    "df_r13[\"Max Expr\"] = np.max(df_r13, axis = 1)\n",
    "keep_rat = list(df_r13[df_r13[\"Max Expr\"] > 5].index)\n",
    "\n",
    "df_m13[\"Max Expr\"] = np.max(df_m13, axis = 1)\n",
    "keep_mouse = list(df_m13[df_m13[\"Max Expr\"] > 5].index)\n",
    "\n",
    "#Restrict to only genes in both\n",
    "keep = list(set(keep_rat + keep_mouse))\n",
    "df_r13 = df_r13.loc[keep].sort_index()\n",
    "df_m13 = df_m13.loc[keep].sort_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9117d946",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copied from above\n",
    "def tau(x):\n",
    "    max_x = max(x)\n",
    "    n = len(x)\n",
    "    return np.sum((1-x/max_x)/(n-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f7f239",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove cerebellum as it is highly similar to forebrain\n",
    "df_r13 = df_r13.drop([\"Cerebellum\"], axis = 1)\n",
    "df_m13 = df_m13.drop([\"Cerebellum\"], axis = 1)\n",
    "df_lfc = np.log2((df_m13 + 1)/(df_r13 + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d8a2332",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute tau\n",
    "df_r13[\"Tau\"] = df_r13.apply(tau, axis = 1)\n",
    "df_m13[\"Tau\"] = df_m13.apply(tau, axis = 1)\n",
    "df_kaes_tau = pd.DataFrame()\n",
    "df_kaes_tau[\"Tau rat\"] = df_r13[\"Tau\"]\n",
    "df_kaes_tau[\"Tau mouse\"] = df_m13[\"Tau\"]\n",
    "\n",
    "#Take average in mouse and rat\n",
    "df_kaes_tau[\"Mean tau\"] = np.mean(df_kaes_tau, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81778e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute variance across tissues\n",
    "df_lfc[\"Variance\"] = np.var(df_lfc, axis = 1)\n",
    "df_lfc_var = df_lfc[[\"Variance\"]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40269af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "out = []\n",
    "for index, row in df_kaes_tau.iterrows():\n",
    "    try:\n",
    "        out.append([d[index]] + list(row))\n",
    "    except:\n",
    "        pass\n",
    "df_kaes_tau = pd.DataFrame(out)\n",
    "df_kaes_tau.columns = [\"Gene\", \"Tau rat\", \"Tau mouse\", \"Mean tau\"]\n",
    "df_kaes_tau = df_kaes_tau.set_index(\"Gene\")[[\"Mean tau\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "166b3013",
   "metadata": {},
   "outputs": [],
   "source": [
    "out = []\n",
    "for index, row in df_lfc_var.iterrows():\n",
    "    try:\n",
    "        out.append([d[index]] + list(row))\n",
    "    except:\n",
    "        pass\n",
    "df_lfc_var = pd.DataFrame(out)\n",
    "df_lfc_var.columns = [\"Gene\", \"LFC variance\"]\n",
    "df_lfc_var = df_lfc_var.set_index(\"Gene\")[[\"LFC variance\"]]\n",
    "df_lfc_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "610711fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make files to run multiple linear regression\n",
    "try:\n",
    "    os.mkdir(\"Final/RegInput\")\n",
    "except:\n",
    "    print(\"Dir exists\")\n",
    "for file in os.listdir(\"Final/Div/\"):\n",
    "    #Read in the info for the cell type\n",
    "    v = pd.read_csv(\"Final/Div/\" + file, sep = \",\").set_index(\"Unnamed: 0\")\n",
    "    \n",
    "    #Read in constraint\n",
    "    cons = pd.read_csv(\"Final/Pritchard_Constraint_Metric_GeneName.txt\", sep = \"\\t\").drop_duplicates(\"Gene name\").set_index(\"Gene name\")\n",
    "    cons = cons[[\"post_mean\"]]\n",
    "    \n",
    "    #Get the right sample for brain and connective tissue\n",
    "    if \"Brain\" in file:\n",
    "        v[\"Total expression\"] = (v[\"MR1_ri Norm CPM\"] + v[\"RM1_mi Norm CPM\"])/2\n",
    "    else:\n",
    "        v[\"Total expression\"] = (v[\"MR1_ri Norm CPM\"] + v[\"RM2_mi Norm CPM\"])/2\n",
    "    \n",
    "    #Restrict to only columns of interest\n",
    "    v = v[[\"Proportion intrinsic\", \"Proportion extrinsic\", \"Proportion interaction\", \"Intrinsic\", \"Extrinsic\", \"Interaction\", \"Total expression\", \"HM/HR\"]]\n",
    "    out = []\n",
    "    for index, row in v.iterrows():\n",
    "        try:\n",
    "            out.append([d[index]] + list(row))\n",
    "        except:\n",
    "            pass\n",
    "    \n",
    "    #Compute absolute log-fold change after makking data frame\n",
    "    df = pd.DataFrame(out).set_index(0)\n",
    "    df.columns = [\"Proportion intrinsic\", \"Proportion extrinsic\", \"Proportion interaction\", \"Intrinsic\", \"Extrinsic\", \"Interaction\", \"Total expression\", \"HM/HR\"]\n",
    "    df[\"HM/HR\"] = np.abs(df[\"HM/HR\"])\n",
    "    \n",
    "    #Join the various dataframes, require an entry for a gene in all of them\n",
    "    df = df.join(cons).dropna().join(cpg).dropna().join(rns).dropna().join(TAU).dropna().join(df_kaes_tau).dropna().join(df_lfc_var).dropna()\n",
    "    df.columns = [\"Proportion intrinsic\", \"Proportion extrinsic\", \"Proportion interaction\", \"Intrinsic\", \"Extrinsic\", \"Interaction\", \"Total expression\", \"HM/HR\"] + [\"Constraint on expression\", \"Number of CREs\", \"cis-neighborhood size\", \"Tau us\", \"Tau kaes\", \"LFC variance kaes\"]\n",
    "    #Compute CRE density\n",
    "    df[\"CRE density\"] = df[\"Number of CREs\"]/np.minimum(100000, df[\"cis-neighborhood size\"])\n",
    "    df[\"Gene\"] = df.index\n",
    "    df = df.drop_duplicates([\"Gene\"])\n",
    "    df = df.drop([\"Gene\"], axis = 1)\n",
    "    #Output\n",
    "    df.to_csv(\"Final/RegInput/\" + file.replace(\".csv\", \"_LinReg.csv\"), index = True, header = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc356acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute the spearman correlations with our predictors\n",
    "#Also create current Figure 3D\n",
    "#Can change to extrinsic or interaction\n",
    "directory = \"Final/RegInput/\"\n",
    "out = []\n",
    "for file in os.listdir(directory):\n",
    "    #Compute a mess of spearman correlations\n",
    "    v = pd.read_csv(directory + file, sep = \",\")\n",
    "    v[\"CRE density\"] = v[\"Number of CREs\"]/np.maximum(10000, np.minimum(100000, v[\"cis-neighborhood size\"]))\n",
    "    te = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"Total expression\"])\n",
    "    lfc = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"HM/HR\"])\n",
    "    ce = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"Constraint on expression\"])\n",
    "    ncres = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"Number of CREs\"])\n",
    "    cn_size = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"cis-neighborhood size\"])\n",
    "    tau_ = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"Tau us\"])\n",
    "    reg_dens = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"CRE density\"])\n",
    "    tau_kaes = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"Tau kaes\"])\n",
    "    lfc_var = spearmanr(v[\"Proportion intrinsic\"]-v[\"Proportion extrinsic\"], v[\"LFC variance kaes\"])\n",
    "    out.append([d_ct_abrev[file_to_celltype(file)], te[0], te[1], lfc[0], lfc[1], ce[0], ce[1], ncres[0], ncres[1], cn_size[0], cn_size[1], tau_[0], tau_[1], reg_dens[0], reg_dens[1], tau_kaes[0], tau_kaes[1], lfc_var[0], lfc_var[1]])\n",
    "df = pd.DataFrame(out)\n",
    "\n",
    "#Create dataframe of the spearman correlations\n",
    "df.columns = [\"Cell type\", \"Total expression rho\", \"Total expression p-value\", \"LFC rho\", \"LFC p-value\", \"Constraint rho\", \"Constraint p-value\", \"Number CREs rho\", \"Number CREs p-value\", \"Nearest gene dist rho\", \"Nearest gene dist p-value\", \"Tau us rho\", \"Tau us p-value\", \"CRE density rho\", \"CRE density p-value\", \"Tau kaes rho\", \"Tau kaes p-value\", \"LFC var kaes rho\", \"LFC var kaes p-value\"]\n",
    "df.to_csv(\"Final/Spearman_Prop_Intrinsic-Extrinsic_Predictors.csv\", sep = \",\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c6d4e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make illustrative figures\n",
    "#Colors\n",
    "mouse = \"#F2C911\"\n",
    "rat = \"#65B0AC\"\n",
    "intrinsic = \"#F55F00\"\n",
    "extrinsic = \"#7D9AF4\"\n",
    "reinforcing = \"#9B00F5\"\n",
    "opposing = \"#F50901\"\n",
    "interaction = \"#1E771A\"\n",
    "\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")\n",
    "d_ct_abrev = {\"Forebrain glutamatergic progenitors\":\"brain.glut.prog\", \"Forebrain GABAergic progenitors\":\"brain.GABA.prog\", \"Intermediate progenitors\":\"inter.prog\", \"Forebrain glutamatergic neurons\":\"brain.glut.neu\", \"Forebrain GABAergic neurons\":\"brain.GABA.neu\", \"Spinal GABAergic neurons\":\"spine.GABA.neu\", \"Spinal glutamatergic neurons\":\"spine.glut.neu\", \"Chondrocyte\":\"chondrocyte\", \"Chondrocytes\":\"chondrocyte\", \"Mesenchyme 0\":\"mesen.0\", \"Mesenchyme 2\":\"mesen.2\", \"Mesenchyme cycling\":\"mesen.cyc\", \"Forebrain GABAergic neurons 0\":\"Toss\"}\n",
    "\n",
    "#Function to convert file to cell type\n",
    "def file_to_celltype(x):\n",
    "    x = x.split(\"/\")[-1]\n",
    "    x = x.replace(\"_GO_Molecular_Function_2023\", \"\").replace(\"_GO_Biological_Process_2023\", \"\").replace(\"Mesechyme\", \"Mesenchyme\").replace(\"Brain_Or_2010_Div_New4_NewNorm_\", \"\")\n",
    "    x = x.replace(\"Mesenchymal_Or_2010_Div_New4_NewNorm_\", \"\").replace(\"Chondrocyte_Or_2010_Div_New4_NewNorm_\", \"\").replace(\".csv\", \"\").replace(\"_\", \" \").replace(\" all\", \"\")\n",
    "    x = x.replace(\"Glutamatergic\", \"Forebrain glutamatergic\").replace(\"GABAergic\", \"Forebrain GABAergic\").replace(\"Spinal Forebrain\", \"Spinal\").replace(\"Chondrocytes\", \"Chondrocyte\")\n",
    "    return x.replace(\" LinReg\", \"\")\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52ec3b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make versions of expression plots with made up data\n",
    "#mouse_vals is the mouse expression and rat_vals is the rat expression\n",
    "def ill_poss(mouse_vals, rat_vals, title, xlabs=[\"E11.5\", \"E13.5\", \"E15.5\"]):\n",
    "    #Create scatterplots\n",
    "    plt.scatter(x = xlabs, y = mouse_vals, color = mouse, s = 100, alpha = 1)\n",
    "    plt.scatter(x = xlabs, y = rat_vals, color = rat, s = 100, alpha = 1)\n",
    "    plt.title(title)\n",
    "    plt.legend({\"Mouse\":mouse, \"Rat\":rat})\n",
    "    plt.show()\n",
    "    tp = pd.DataFrame()\n",
    "    \n",
    "    #Add barplots for mean values\n",
    "    tp[\"Chimera (Donor-Host)\"] = [\"Rat-like env.\", \"Rat-like env.\", \"Mouse-like env.\", \"Mouse-like env.\", \"Host\", \"Host\"]\n",
    "    tp[\"Cell species\"] = [\"Mouse\", \"Rat\", \"Mouse\", \"Rat\", \"Mouse\", \"Rat\"]\n",
    "    tp[\"Norm CPM\"] = [mouse_vals[2], rat_vals[2], mouse_vals[1], rat_vals[1], mouse_vals[1], rat_vals[2]]\n",
    "    sns.set(font_scale=1.5)\n",
    "    sns.set_style(\"white\")\n",
    "    sns.barplot(data = tp, y = \"Norm CPM\", x = \"Chimera (Donor-Host)\", hue = \"Cell species\", palette = {\"Mouse\":mouse, \"Rat\":rat}, alpha = 1, order = [\"Host\", \"Rat-like env.\", \"Mouse-like env.\"])\n",
    "    plt.ylabel(\"Pseudobulked normalized counts\")\n",
    "    plt.xlabel(\"Extrinsic environment\")\n",
    "    plt.title(title)\n",
    "    #plt.legend(bbox_to_anchor=(1.375, 1.05))\n",
    "    plt.legend([],[], frameon=False)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a652f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#An option for opposing intrinsic/extrinsic\n",
    "ill_poss([10, 20, 30], [5, 12.5, 20], \"Intrinsically higher mouse and increasing expression\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80280592",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Another opposing option\n",
    "ill_poss([15, 10, 5], [30, 20, 10], \"Intrinsically higher rat and decreasing expression\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687b7e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A reinforcing option\n",
    "ill_poss([5, 12.5, 20], [10, 20, 30], \"Intrinsically higher rat and increasing expression\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d9ef263",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Another reinforcing option\n",
    "ill_poss([30, 20, 10], [15, 10, 5], \"Intrinsically higher mouse and decreasing expression\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adaf7be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A way to get \"extrinsic divergence\" with only intrinsic divergence\n",
    "ill_poss([30, 20, 10], [29.5, 20.5, 9.5], \"Decreasing expression and heterochronic shift\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58f167da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Another way to get \"extrinsic divergence\" with only intrinsic divergence\n",
    "ill_poss([10, 20, 30], [10.5, 19.5, 30.5], \"Increasing expression and heterochronic shift\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48b66aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Commented and uncommented things below so that the comments before them match\n",
    "#E.g. if \"For interaction\" is what is uncommented, it should be what is uncommented for everthing\n",
    "\n",
    "#For opposing/reinforcing\n",
    "#folder = \"GSEAPY_New4_ReinfOppo_SameDirSigned\"\n",
    "#sort_by = \"Multiply\"\n",
    "\n",
    "#For pure extrinsic\n",
    "folder = \"GSEAPY_New4_Extrinsic_SameDirSameExpr\"\n",
    "sort_by = \"Signed proportion extrinsic\"\n",
    "\n",
    "#For interaction\n",
    "folder = \"GSEAPY_New4_Interaction_DifTimePatternSigned\"\n",
    "sort_by = \"Signed interaction\"\n",
    "\n",
    "fdr = []\n",
    "enrichment = []\n",
    "ct_list = []\n",
    "term_list = []\n",
    "es = []\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    #Only do for brain since that is what we have Kaessmann data for\n",
    "    if \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        ct = file.replace(\"Brain_Or_2010_Div_New4_NewNorm_\", \"\").replace(\"Mesenchymal_Or_2010_Div_New4_NewNorm_\", \"\").replace(\"Chondrocyte_Or_2010_Div_New4_NewNorm_\", \"\").replace(\".csv\", \"\")\n",
    "        ct = d_ct_abrev[file_to_celltype(file)].replace(\".\", \"_\")\n",
    "        \n",
    "        #Get results of enrichment\n",
    "        vv = pd.read_csv(\"Final/\" + folder + \"/\" + file.replace(\".csv\", \"\") + \"/gseapy.gene_set.prerank.report.csv\", sep = \",\")\n",
    "        \n",
    "        #For reinforcing/opposing\n",
    "        #v[sort_by] = np.sign(v[\"Intrinsic\"])*v[\"Proportion intrinsic\"]*v[\"Proportion extrinsic\"]*np.sign(v[\"Extrinsic\"])\n",
    "        \n",
    "        #For pure extrinsic\n",
    "        #v[\"Signed proportion extrinsic\"] = np.sign(v[\"Extrinsic\"])*v[\"Proportion extrinsic\"]\n",
    "        \n",
    "        #For signed interaction\n",
    "        v[\"Signed interaction\"] = np.sign(v[\"Interaction\"])*v[\"Proportion interaction\"]\n",
    "        \n",
    "        v = v[[\"Unnamed: 0\", sort_by]]\n",
    "        v.columns = [\"Gene\", sort_by]\n",
    "        \n",
    "        #Add terms\n",
    "        for index, row in vv.iterrows():\n",
    "            term = row[\"Term\"]\n",
    "            term_list.append(row[\"Term\"])\n",
    "            #For opposing/reinforcing\n",
    "            #gene_set = d_OR[term]\n",
    "            \n",
    "            #For pure extrinsic\n",
    "            #gene_set = d_SameDir[term]\n",
    "            \n",
    "            #For interaction\n",
    "            gene_set = d_DifDir[term]\n",
    "            genes = row[\"Lead_genes\"].split(\";\")\n",
    "            fdr.append(max(row[\"FDR q-val\"], 0.001))\n",
    "            es.append(row[\"ES\"])\n",
    "            if row[\"ES\"] < 0:\n",
    "                asc = True\n",
    "            else:\n",
    "                asc = False\n",
    "            v = v.sort_values(sort_by, ascending = asc)\n",
    "            \n",
    "            #Find the gene rank that was used as a cutoff in GSEAPY\n",
    "            cut = 0\n",
    "            for index, row in v.iterrows():\n",
    "                cut += 1\n",
    "                if genes[-1] == row[\"Gene\"]:\n",
    "                    break\n",
    "        \n",
    "            #Using that rank, get the number of genes that drive the enrichment and the number of genes going against it at an equivalent cutoff\n",
    "            genes_agree = len(np.intersect1d(list(v[\"Gene\"])[:cut], gene_set))\n",
    "            genes_disagree = len(np.intersect1d(list(v[\"Gene\"])[::-1][:cut], gene_set))\n",
    "            ct_list.append(d_ct_abrev[file_to_celltype(file)])\n",
    "            \n",
    "            #Make sign of the fold-change properly match the sign of the original GSEAPY enrichment\n",
    "            if asc:\n",
    "                enrichment.append(-np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "            else:\n",
    "                enrichment.append(np.log2((genes_agree + 1)/(genes_disagree + 1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd485af7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create dataframe of cell types, terms, log fold-enrichments, and FDRs\n",
    "df_plot_new = pd.DataFrame([term_list, ct_list, enrichment, fdr, es]).T\n",
    "df_plot_new.columns = [\"Gene set\", \"Cell type\", \"Log$_{2}$ fold-enrichment\", \"FDR\", \"ES\"]\n",
    "df_plot_new[\"-Log$_{10}$ FDR\"] = -np.log10(df_plot_new[\"FDR\"].astype(float))\n",
    "df_plot_new[\"Log$_{2}$ fold-enrichment\"] = df_plot_new[\"Log$_{2}$ fold-enrichment\"].astype(float)\n",
    "df_plot_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "136464c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subplots\n",
    "sns.set(font_scale = 1.5)\n",
    "sns.set_style(\"white\")\n",
    "fig, ax = plt.subplots(sharex=True, sharey=True)\n",
    "\n",
    "# Separate data by color\n",
    "grey_data = df_plot_new[df_plot_new['Gene set'] != 'UpRat_DownMouse']\n",
    "blue_data = df_plot_new[df_plot_new['Gene set'] == 'UpRat_DownMouse']\n",
    "\n",
    "# Scatter plots with zorder\n",
    "sns.scatterplot(x=grey_data['Log$_{2}$ fold-enrichment'], y=grey_data['-Log$_{10}$ FDR'], c='orange', label='Increasing in mouse and\\ndecreasing in rat', zorder=1)\n",
    "sns.scatterplot(x=blue_data['Log$_{2}$ fold-enrichment'], y=blue_data['-Log$_{10}$ FDR'], c='blue', label='Increasing in rat and\\ndecreasing in mouse', zorder=2)\n",
    "\n",
    "# Customize plot settings\n",
    "#plt.title(\"Enrichments split by reinforcing and opposing\")\n",
    "plt.title(\"Enrichments split by directionality of\\nchange in expression over time\")\n",
    "ax.set_xlabel('Log$_{2}$ fold-enrichment')\n",
    "ax.set_ylabel('-Log$_{10}$ FDR')\n",
    "\n",
    "# Repeat similar steps for other subplots (7 Day, 14 Day, 28 Day)\n",
    "\n",
    "# Show the legend\n",
    "ax.legend(bbox_to_anchor= (2, 1))\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b8204a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### The below three cells are repeats of the third cell above this one, so are not commented ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4102ce38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting the results of Enrichment analysis with absolute measures\n",
    "#For Reinforcing/Opposing\n",
    "enrich = []\n",
    "fdr = []\n",
    "cell_types = []\n",
    "genes = []\n",
    "for file in os.listdir(\"Final/GSEAPY_New4_ReinfOppo_SameDir_Abs\"):\n",
    "    if \"neurons_0\" not in file:\n",
    "        v = pd.read_csv(\"Final/GSEAPY_New4_ReinfOppo_SameDir_Abs/\" + file + \"/gseapy.gene_set.prerank.report.csv\", sep = \",\")\n",
    "        out = []\n",
    "        for index, row in v.iterrows():\n",
    "            enrich.append(row[\"ES\"])\n",
    "            fdr.append(row[\"FDR q-val\"])\n",
    "            cell_types.append(d_ct_abrev[file_to_celltype(file)].replace(\" \", \".\"))\n",
    "            genes.append(row[\"Lead_genes\"])\n",
    "d_en = {}\n",
    "for i in range(len(cell_types)):\n",
    "    d_en[cell_types[i]] = genes[i]\n",
    "    \n",
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "gene_set = df_same.index\n",
    "\n",
    "ct = []\n",
    "enrichment = []\n",
    "asc = False\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        v[\"Multiply\"] = np.abs(v[\"Proportion intrinsic\"]*v[\"Proportion extrinsic\"])\n",
    "        asc = False\n",
    "        v = v.sort_values(\"Multiply\", ascending = asc)\n",
    "        genes = d_en[d_ct_abrev[file_to_celltype(file)]].split(\";\")\n",
    "        cut = 0\n",
    "        for index, row in v.iterrows():\n",
    "            cut += 1\n",
    "            if genes[-1] == row[\"Unnamed: 0\"]:\n",
    "                break\n",
    "        genes_agree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[:cut], gene_set))\n",
    "        genes_disagree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[::-1][:cut], gene_set))\n",
    "        ct.append(d_ct_abrev[file_to_celltype(file)])\n",
    "        if asc:\n",
    "            enrichment.append(-np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "        else:\n",
    "            enrichment.append(np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "            \n",
    "sns.set(font_scale = 1.5)\n",
    "sns.set_style(\"white\")\n",
    "to_plot = pd.DataFrame([enrichment, fdr, cell_types]).T\n",
    "to_plot.columns = [\"Log$_2$ fold-enrichment\", \"FDR\", \"Cell type\"]\n",
    "sns.barplot(data = to_plot, x = \"Cell type\", y = \"Log$_2$ fold-enrichment\", color = \"grey\")\n",
    "plt.title(\"Enrichment for reinforcing/opposing\\nand same direction over time genes\")\n",
    "#plt.legend(bbox_to_anchor=(1.5, 1.05))\n",
    "plt.legend([],[], frameon=False)\n",
    "plt.xticks(rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f174db58",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting the results of Enrichment analysis with absolute measures\n",
    "#For Interaction/differences in temporal pattern of gene expression\n",
    "enrich = []\n",
    "fdr = []\n",
    "cell_types = []\n",
    "genes = []\n",
    "for file in os.listdir(\"Final/GSEAPY_New4_Interaction_DifTimePattern\"):\n",
    "    if \"neurons_0\" not in file:\n",
    "        v = pd.read_csv(\"Final/GSEAPY_New4_Interaction_DifTimePattern/\" + file + \"/gseapy.gene_set.prerank.report.csv\", sep = \",\")\n",
    "        out = []\n",
    "        for index, row in v.iterrows():\n",
    "            enrich.append(row[\"ES\"])\n",
    "            fdr.append(row[\"FDR q-val\"])\n",
    "            cell_types.append(d_ct_abrev[file_to_celltype(file)].replace(\" \", \".\"))\n",
    "            genes.append(row[\"Lead_genes\"])\n",
    "d_en = {}\n",
    "for i in range(len(cell_types)):\n",
    "    d_en[cell_types[i]] = genes[i]\n",
    "    \n",
    "df_oppo = df[df[\"Mult\"] < -0.25]\n",
    "gene_set = df_oppo.index\n",
    "\n",
    "ct = []\n",
    "enrichment = []\n",
    "asc = False\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        asc = False\n",
    "        v = v.sort_values(\"Proportion interaction\", ascending = asc)\n",
    "        genes = d_en[d_ct_abrev[file_to_celltype(file)]].split(\";\")\n",
    "        cut = 0\n",
    "        for index, row in v.iterrows():\n",
    "            cut += 1\n",
    "            if genes[-1] == row[\"Unnamed: 0\"]:\n",
    "                break\n",
    "        genes_agree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[:cut], gene_set))\n",
    "        genes_disagree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[::-1][:cut], gene_set))\n",
    "        print(genes_agree, genes_disagree, d_ct_abrev[file_to_celltype(file)])\n",
    "        ct.append(d_ct_abrev[file_to_celltype(file)])\n",
    "        if asc:\n",
    "            enrichment.append(-np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "        else:\n",
    "            enrichment.append(np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "            \n",
    "sns.set(font_scale = 1.5)\n",
    "sns.set_style(\"white\")\n",
    "to_plot = pd.DataFrame([enrichment, fdr, cell_types]).T\n",
    "to_plot.columns = [\"Log$_2$ fold-enrichment\", \"FDR\", \"Cell type\"]\n",
    "sns.barplot(data = to_plot, x = \"Cell type\", y = \"Log$_2$ fold-enrichment\", color = \"grey\")\n",
    "plt.title(\"Enrichment for interaction\\nand opposite direction over time genes\")\n",
    "#plt.legend(bbox_to_anchor=(1.5, 1.05))\n",
    "plt.legend([],[], frameon=False)\n",
    "plt.xticks(rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549eba70",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plotting the results of Enrichment analysis with absolute measures\n",
    "#For pure extrinsic everything\n",
    "\n",
    "enrich = []\n",
    "fdr = []\n",
    "cell_types = []\n",
    "genes = []\n",
    "for file in os.listdir(\"Final/GSEAPY_New4_Extrinsic_SameDir\"):\n",
    "    if \"neurons_0\" not in file:\n",
    "        v = pd.read_csv(\"Final/GSEAPY_New4_Extrinsic_SameDir/\" + file + \"/gseapy.gene_set.prerank.report.csv\", sep = \",\")\n",
    "        out = []\n",
    "        for index, row in v.iterrows():\n",
    "            enrich.append(row[\"ES\"])\n",
    "            fdr.append(row[\"FDR q-val\"])\n",
    "            cell_types.append(d_ct_abrev[file_to_celltype(file)].replace(\" \", \".\"))\n",
    "            genes.append(row[\"Lead_genes\"])\n",
    "d_en = {}\n",
    "for i in range(len(cell_types)):\n",
    "    d_en[cell_types[i]] = genes[i]\n",
    "    \n",
    "df_same = df[df[\"Mult\"] > 0.7]\n",
    "gene_set = df_oppo.index\n",
    "\n",
    "ct = []\n",
    "enrichment = []\n",
    "asc = False\n",
    "for file in os.listdir(\"Final/Div\"):\n",
    "    if \"Brain\" in file:\n",
    "        v = pd.read_csv(\"Final/Div/\" + file, sep = \",\")\n",
    "        asc = False\n",
    "        v = v.sort_values(\"Proportion extrinsic\", ascending = asc)\n",
    "        genes = d_en[d_ct_abrev[file_to_celltype(file)]].split(\";\")\n",
    "        cut = 0\n",
    "        for index, row in v.iterrows():\n",
    "            cut += 1\n",
    "            if genes[-1] == row[\"Unnamed: 0\"]:\n",
    "                break\n",
    "        genes_agree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[:cut], gene_set))\n",
    "        genes_disagree = len(np.intersect1d(list(v[\"Unnamed: 0\"])[::-1][:cut], gene_set))\n",
    "        print(genes_agree, genes_disagree, d_ct_abrev[file_to_celltype(file)])\n",
    "        ct.append(d_ct_abrev[file_to_celltype(file)])\n",
    "        if asc:\n",
    "            enrichment.append(-np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "        else:\n",
    "            enrichment.append(np.log2((genes_agree + 1)/(genes_disagree + 1)))\n",
    "            \n",
    "sns.set(font_scale = 1.5)\n",
    "sns.set_style(\"white\")\n",
    "to_plot = pd.DataFrame([enrichment, fdr, cell_types]).T\n",
    "to_plot.columns = [\"Log$_2$ fold-enrichment\", \"FDR\", \"Cell type\"]\n",
    "sns.barplot(data = to_plot, x = \"Cell type\", y = \"Log$_2$ fold-enrichment\", color = \"grey\")\n",
    "plt.title(\"Enrichment for extrinsic\\nand same direction over time genes\")\n",
    "#plt.legend(bbox_to_anchor=(1.5, 1.05))\n",
    "plt.legend([],[], frameon=False)\n",
    "plt.xticks(rotation=90)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
